### PR TITLE
[#2533,#3824] Add TCP keepalive option configs (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #2533, #3824

Unit tests passed, running core tests.

I will be playing with `tc` tomorrow to see if I can demonstrate any effect from changing these values. I have demonstrated that setting `tcp_keepalive_time` to 0 results in an `errno` of 22, which is "Invalid argument". Not very useful. :)

### CODE REVIEW

File: schemas/configuration/v4/service_account_environment.json.in
Comment (korydraughn) [inline schemas/configuration/v4/service_account_environment.json.in:43]: Just noting the docs will need to be updated to include descriptions for these too.

**Review by korydraughn — COMMENTED** [review]
Always happy to see code consolidation.

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp:57]: Yes, this is a re-implementation of `try_twice_to_create_socket` with a variable number of retries. I'm willing to try out `with_durability` if that seems better.

Originally, I had the socket options being set as part of this new implementation which is partially why there even is a new implementation. But I thought better of it and left the socket option setting to the caller(s).

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp:354]: I cannot find any information about the range of acceptable values for these, so if anybody knows, that would be nice.

The main thing that I have discovered is that 0 is not valid for `TCP_KEEPIDLE`. I suspect that this may be the case for `TCP_KEEPINTVL` and `TCP_KEEPCNT`, but I haven't tried it yet.

I went back and forth on whether non-negative integers was the right domain, but this seems like a strong case for limiting to positive integers.

File: lib/core/src/sockComm.cpp
Comment (trel) [reply lib/core/src/sockComm.cpp:57]: `with_durability` seems... the right tool here.

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp:817]: Correct

File: lib/core/src/sockComm.cpp
Comment (trel) [reply lib/core/src/sockComm.cpp:354]: there's a spec / RFC somewhere...

File: schemas/configuration/v4/service_account_environment.json.in
Comment (alanking) [reply schemas/configuration/v4/service_account_environment.json.in:43]: Right, forgot to hit submit on that PR... https://github.com/irods/irods_docs/pull/209

File: lib/core/src/getRodsEnv.cpp
Comment (alanking) [reply lib/core/src/getRodsEnv.cpp:359]: Whups

**Comment by alanking** [comment]
Will address clang-format and clang-tidy. Also, commit message needs to be updated as we moved away from the server_config options to client environment configuration.

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp:354]: Alternatively, we could just try to apply whatever setting the user has provided and let it fail with an error. This is, essentially, exposing the TCP keepalive options for iRODS connections as configurable options, so we could just fall back on the behavior of `setsockopt` instead of worrying about it. Thoughts?

Comment (korydraughn) [reply lib/core/src/sockComm.cpp:354]: Yes. As long as we're confident we're holding the OS API calls correctly, I think it's fine. No need to test OS APIs.

Comment (alanking) [reply lib/core/src/sockComm.cpp:354]: I think so. This implementation is pulled almost directly from the example here: https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/programming.html

I will leave a lengthy comment here in the code explaining why we don't check the values and some words in the documentation for these options warning the user to be careful.

Comment (korydraughn) [reply lib/core/src/sockComm.cpp:354]: Sounds good.

Just so I understand, the logic you have is not going to change, correct?
You're going to add comments to help the next developer only.

Comment (korydraughn) [reply lib/core/src/sockComm.cpp:354]: I see. And if they set it to a bad value `setsockopt` will report an error and the client will see that in the iRODS error code.

Carry on.

Comment (alanking) [reply lib/core/src/sockComm.cpp:354]: Exactly

Comment (alanking) [reply lib/core/src/sockComm.cpp:354]: Oops... It turns out that we can't do this approach because we need to set the members of the `rodsEnv` to *something* to indicate that they are not set, and I was using -1 for this purpose. I have confirmed that 0 is an invalid input for all 3 of these socket options, so we should only set the socket option when the value is a positive integer.

Also, I think that the default values in the documentation should be -1 so that "unset" is the default, but we will note what the defaults from the TCP spec are.

Thoughts?

Comment (trel) [reply lib/core/src/sockComm.cpp:354]: seems fair all around.

Comment (korydraughn) [reply lib/core/src/sockComm.cpp:354]: No objections here.

Comment (alanking) [reply lib/core/src/sockComm.cpp]: Done. Added words to docs to this effect, too

File: lib/core/src/getRodsEnv.cpp
Comment (alanking) [reply lib/core/src/getRodsEnv.cpp]: Done

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp]: I just used the passed-in `_retries` alongside the default `durability_options` (1000ms delay with multiplier of 1). Should we expose the other options as parameters of the function? I feel like the defaults are okay.

**Comment by alanking** [comment]
Ignoring clang-format's insistence on indenting the anonymous namespace.

I may need to revisit the `with_durability` usage. Let me know if we should use a different set of options.

Unit tests passed with current changes

File: lib/core/src/sockComm.cpp
Comment (korydraughn) [reply lib/core/src/sockComm.cpp]: I agree. The defaults are fine for now.

**Comment by korydraughn** [comment]
Your use of `with_durability` looks fine to me. What part of it raises concerns?

File: lib/core/src/sockComm.cpp
Comment (alanking) [reply lib/core/src/sockComm.cpp]: Great, resolving!

**Comment by alanking** [comment]
Just wanted to make sure we were okay with the options being used. If it looks okay, then I'm okay with it

**Comment by korydraughn** [comment]
I think setting only the number of retries is a perfectly good starting point. I'm happy with the solution presented.

**Comment by trel** [comment]
agreed.  ship it.   keepalive will surely never die.  we'll hear about it again at some point.

**Comment by korydraughn** [comment]
I'm wondering if we can add a unit test that proves the rcConnect call fails with the appropriate socket error?

We can set environment variables via `setenv`. And having a test that at least shows the iRODS libs honor the documentation you've written in irods_docs is good.

Thoughts?

**Comment by alanking** [comment]
I think that's a fine idea. I'll add at least one...

**Comment by alanking** [comment]
Core tests passed. Still need to add a new test or three

**Comment by alanking** [comment]
Force-push due to rebase. Added one test for each new config, all of which pass.

File: scripts/irods/test/test_misc.py
Comment (korydraughn) [inline scripts/irods/test/test_misc.py:405]: Consider adding more of the error message to the expected output value to guard against false positives in the future. While unlikely to happen, if `-1790220` became a valid error code and was returned here, the test would accept it.

I'm assuming this actually returns something like `-179022 SOCK_ERROR`.

_Same thing for other tests._

**Review by korydraughn — COMMENTED** [review]
Tests look good. Just one comment.

File: scripts/irods/test/test_misc.py
Comment (alanking) [reply scripts/irods/test/test_misc.py:405]: Done

**Comment by korydraughn** [comment]
Agreed. If you can't provide a test that demonstrates the timeouts without taking 2 hours, then don't worry about it. We can capture that in an issue and add one later once we learn more.

You may have to adjust some OS settings to prove the timeout case. If you choose to investigate that, consider doing that in a docker container to avoid screwing up the host OS.

**Comment by korydraughn** [comment]
Are you planning on adding anything else to this?

**Comment by alanking** [comment]
I hadn't planned on it, no. I can try the timeout thing tonight if that seems like something we want to do before merging. Don't want to hold up my machine for that long ;)

**Comment by korydraughn** [comment]
Just checking. Approving now.

**Review by korydraughn — COMMENTED** [review]
If the tests pass and you're happy with it, please squash to taste and leave off the pounds.

**Comment by alanking** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
Ship it

**Pull request merged**
